be/src/exec/stream/scan/stream_scan_operator.cpp

```diff
@@ -32,6 +32,126 @@ StreamScanOperator::StreamScanOperator(OperatorFactory* factory, int32_t id, int
         : ConnectorScanOperator(factory, id, driver_sequence, dop, scan_node),
           _is_stream_pipeline(is_stream_pipeline) {}
 
+StreamScanOperator::~StreamScanOperator() {
+    auto* state = runtime_state();
+    if (state == nullptr) {
+        return;
+    }
+
+    while (!_closed_chunk_sources.empty()) {
+        ChunkSourcePtr chunk_source_ptr = _closed_chunk_sources.front();
+        _closed_chunk_sources.pop();
+        std::lock_guard guard(_task_mutex);
+        chunk_source_ptr->close(state);
+        chunk_source_ptr = nullptr;
+    }
+
+    // defensive programming,_chunk_sources should have been closed normally
+    for (size_t i = 0; i < _chunk_sources.size(); i++) {
+        if (_chunk_sources[i] != nullptr) {
+            _chunk_sources[i]->close(state);
+            _chunk_sources[i] = nullptr;
+            detach_chunk_source(i);
+        }
+    }
+}
+
+void StreamScanOperator::_reset_chunk_source(RuntimeState* state, int chunk_source_index) {
+    auto tablet_id = _chunk_sources[chunk_source_index]->get_lane_owner();
+    auto binlog_offset = _stream_epoch_manager->get_binlog_offset(state->fragment_instance_id(), _id, tablet_id);
+
+    DCHECK(binlog_offset != nullptr);
+    _chunk_sources[chunk_source_index]->set_stream_offset(binlog_offset->tablet_version, binlog_offset->lsn);
+    _chunk_sources[chunk_source_index]->set_epoch_limit(_current_epoch_info.max_scan_rows,
+                                                        _current_epoch_info.max_exec_millis);
+    _chunk_sources[chunk_source_index]->reset_status();
+}
+
+// In order to not modify the code of ScanOperator,
+// there is lots of repetition with the method of _pickup_morsel in ScanOperator,
+// maybe can abstract a method contains the same code later
+Status StreamScanOperator::_pickup_morsel(RuntimeState* state, int chunk_source_index) {
+    DCHECK(_morsel_queue != nullptr);
+    _close_chunk_source(state, chunk_source_index);
+
+    // NOTE: attach an active source before really creating it, to avoid the race condition
+    bool need_detach = true;
+    attach_chunk_source(chunk_source_index);
+    DeferOp defer([&]() {
+        if (need_detach) {
+            detach_chunk_source(chunk_source_index);
+        }
+    });
+
+    ASSIGN_OR_RETURN(auto morsel, _morsel_queue->try_get());
+    if (_lane_arbiter != nullptr) {
+        while (morsel != nullptr) {
+            auto [lane_owner, version] = morsel->get_lane_owner_and_version();
+            auto acquire_result = _lane_arbiter->try_acquire_lane(lane_owner);
+            if (acquire_result == query_cache::AR_BUSY) {
```

Comment:
why consider the query_cache? For StreamOperator I think there's no need to consider query cache ?

Reply:
ok